katsdpcontroller/test/test_product_config.py

```diff
@@ -1,73 +1,1012 @@
+            product_config._normalise_output_channels(100, (60, 50))
+
+    def test_overflow(self) -> None:
+        with assert_raises_regex(ValueError,
+                                 r'output_channels \(0:101\) overflows valid range 0:100'):
+            product_config._normalise_output_channels(100, (0, 101))
+        with assert_raises_regex(ValueError,
+                                 r'output_channels \(-1:1\) overflows valid range 0:100'):
+            product_config._normalise_output_channels(100, (-1, 1))
+
+    def test_align(self) -> None:
+        assert_equal(product_config._normalise_output_channels(1000, (299, 301), 100), (200, 400))
+        assert_equal(product_config._normalise_output_channels(1000, (200, 400), 100), (200, 400))
+
+    def test_misalign(self) -> None:
+        with assert_raises_regex(
+                ValueError,
+                r'n_chans \(789\) is not a multiple of required alignment \(100\)'):
+            product_config._normalise_output_channels(789, (100, 200), 100)
+
+
+class TestServiceOverride:
+    """Test :class:`~.ServiceOverride`."""
+
+    def test_from_config(self) -> None:
+        config = {
+            'config': {'hello': 'world'},
+            'taskinfo': {'image': 'test'},
+            'host': 'test.invalid'
+        }
+        override = ServiceOverride.from_config(config)
+        assert_equal(override.config, config['config'])
+        assert_equal(override.taskinfo, config['taskinfo'])
+        assert_equal(override.host, config['host'])
+
+    def test_defaults(self) -> None:
+        override = ServiceOverride.from_config({})
+        assert_equal(override.config, {})
+        assert_equal(override.taskinfo, {})
+        assert_equal(override.host, None)
+
+
+class TestOptions:
+    """Test :class:`~.Options`."""
+
+    def test_from_config(self) -> None:
+        config = {
+            'develop': True,
+            'wrapper': 'http://test.invalid/wrapper.sh',
+            'service_overrides': {
+                'service1': {
+                    'host': 'testhost'
+                }
+            }
+        }
+        options = Options.from_config(config)
+        assert_equal(options.develop, config['develop'])
+        assert_equal(options.wrapper, config['wrapper'])
+        assert_equal(list(options.service_overrides.keys()), ['service1'])
+        assert_equal(options.service_overrides['service1'].host, 'testhost')
+
+    def test_defaults(self) -> None:
+        options = Options.from_config({})
+        assert_equal(options.develop, False)
+        assert_equal(options.wrapper, None)
+        assert_equal(options.service_overrides, {})
+
+
+class TestSimulation:
+    """Test :class:`~.Simulation`."""
+
+    def test_from_config(self) -> None:
+        config = {
+            'sources': [
+                'PKS 1934-63, radec, 19:39:25.03, -63:42:45.7, (200.0 12000.0 -11.11 7.777 -1.231)',
```

Comment:
Nitpick: You have antennas defined at the top of the tests, why not sources as well ?

Reply:
The antennas are used in multiple tests, while these sources are only used here. If they get used elsewhere I'll make them global.

Reply:
Fair enough